ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,713 @@
+                        for completions in completions.generations
+                    ],
+                },
+            )
+    except Exception:
+        span.set_exc_info(*sys.exc_info())
+        integration.metric(span, "incr", "request.error", 1)
+        raise
+    finally:
+        span.finish()
+        integration.metric(span, "dist", "request.duration", span.duration_ns)
+    return completions
+
+
+@with_traced_module
+async def traced_llm_agenerate(langchain, pin, func, instance, args, kwargs):
+    llm_provider = instance._llm_type
+    prompts = args[0]
+    integration = langchain._datadog_integration
+    span = integration.trace(pin, "%s.%s" % (instance.__module__, instance.__class__.__name__))
+    try:
+        if integration.is_pc_sampled_span(span):
+            for idx, prompt in enumerate(prompts):
+                span.set_tag_str("langchain.request.prompts.%d" % idx, integration.trunc(str(prompt)))
+        model = _extract_model_name(instance)
+        if model is not None:
+            span.set_tag_str("langchain.request.model", model)
+        span.set_tag_str("langchain.request.provider", llm_provider)
+        span.set_tag_str("langchain.request.type", "llm")
+        for param, val in getattr(instance, "_identifying_params", {}).items():
+            if isinstance(val, dict):
+                for k, v in val.items():
+                    span.set_tag_str("langchain.request.%s.parameters.%s.%s" % (llm_provider, param, k), str(v))
+            else:
+                span.set_tag_str("langchain.request.%s.parameters.%s" % (llm_provider, param), str(val))
+        span.set_tag_str("langchain.request.api_key", str(_extract_api_key(instance)))
+
+        completions = await func(*args, **kwargs)
+        if isinstance(instance, langchain.llms.OpenAI):
+            _tag_openai_token_usage(span, completions.llm_output)
+            integration.record_usage(span, completions.llm_output)
+
+        for idx, completion in enumerate(completions.generations):
+            if integration.is_pc_sampled_span(span):
+                span.set_tag_str("langchain.response.completions.%d.text" % idx, integration.trunc(completion[0].text))
+            if completion[0].generation_info is not None:
+                span.set_tag_str(
+                    "langchain.response.completions.%d.finish_reason" % idx,
+                    str(completion[0].generation_info.get("finish_reason")),
+                )
+                span.set_tag_str(
+                    "langchain.response.completions.%d.logprobs" % idx,
+                    str(completion[0].generation_info.get("logprobs")),
+                )
+        if integration.is_pc_sampled_log(span):
+            integration.log(
+                span,
+                "info" if span.error == 0 else "error",
+                "sampled %s.%s" % (instance.__module__, instance.__class__.__name__),
+                attrs={
+                    "prompts": prompts,
```

Comment:
Cardinality is a concern, but due to the nature of LLM prompt-completions this will be inevitable. That's why log sampling is disabled by default and 10% when enabled (and must be explicitly configured to have a different rate)